pandas/core/indexing.py

```diff
@@ -669,6 +670,56 @@ def _get_setitem_indexer(self, key):
 
         return self._convert_to_indexer(key, axis=0)
 
+    @final
+    def _maybe_mask_setitem_value(self, indexer, value):
+        """
+        If we have obj.iloc[mask] = arraylike and arraylike has the same
+        length as obj, we treat this as obj.iloc[mask] = arraylike[mask],
+        similar to Series.__setitem__.
+        """
+        # TODO: why do we only do this for loc, not iloc?
+        # ser = pd.Series(range(5))
+        # mask = np.array([True, False, True, False, True])
+        # ser[mask] = ser * 2  # <- works
+        # ser.loc[mask] = ser * 2  # <- works
+        # ser.iloc[mask] = ser * 2  # <- raises
```

Comment:
loc and setitem align, while iloc does not. Hence this should not work for iloc. This was changed for 1.2 I think and was intended